Support for importers plugins (#174)
pySTEPS/pySTEPS: doc/source/developer_guide/importers_plugins.rst

> .. _importers-plugins:

Reviewer: ```suggestion
.. _importer-plugins:
```
A detail, but I'd suggest to rename it with a singular importer...

================================
Create your own importers plugin
================================

> Since version 3.4, pysteps allows the users to add new importers by installing external

Reviewer: ```suggestion
Since version 1.4, pysteps allows the users to add new importers by installing external
```
I suppose this feature will be included in release 1.4.

Author: Oops, I meant 1.4 :laughing:

packages, called plugins, without modifying the pysteps installation.
These plugins need to follow a particular structure (described next) to allow pysteps
to discover and integrate the new importers to the pysteps interface without any user
intervention.

.. contents:: Table of Contents
    :local:
    :depth: 3

How the plugins work?
=====================

When the plugin is installed, it advertises the new importers to other packages (in our
case, pysteps) using the python `entry points specification`_.

> This new importers are automatically discovered every time that the pysteps library is

Reviewer: ```suggestion
These new importers are automatically discovered every time that the pysteps library is
```

imported. The discovered importers are added as attributes to the io.importers module
and also registered to the io.get_method interface without any user intervention.
In addition, since the installation of the plugins does not modify the actual pysteps
installation (i.e., the pysteps sources), the pysteps library can be updated without
reinstalling the plugin.

.. _`entry points specification`: https://packaging.python.org/specifications/entry-points/

Create your own plugin
======================

There are two ways of creating your own plugin.

> One is building the importers plugin package from scratch.

Reviewer: ```suggestion
The first one involves building the importers plugin package from scratch.
```

An easier alternative is using a `Cookiecutter`_ template that easily create the skeleton

> for an importer plugin template.

Reviewer: ```suggestion
for the new importer plugin.
```

> Although the latter alternative is recommended (the cookiecutter template),

Reviewer: ```suggestion
Although the cookiecutter template is recommended,
```

in the next section we describe step-by-step how to create an importers plugin to
explain in detail all the elements needed for the plugin to work correctly.

After the reader is familiar with the plugins elements, the easier alternative
using the cookiecutter template is presented in the
`Cookiecutter pysteps-plugin template`_ section.

.. _Cookiecutter: https://cookiecutter.readthedocs.io

Plugin project structure
------------------------

Let's suppose that we want to add two new importers to pysteps for reading the radar
composites from the "Awesome Bureau of Composites", kindly abbreviated as "abc".
The composites provided by this institution are available in two different

> formats: Netcdf and Grib2. The details of each format is not important for the rest of

Reviewer: ```suggestion
formats: Netcdf and Grib2. The details of each format are not important for the rest of
```

needed for the plugin to be installed and to work correctly.

The minimal python package to implement an importer plugin have the following
structure:

::

    pysteps-importer-abc        (project name)
    ├── pysteps_importer_abc    (package name)
    │  ├── importer_abc_xyz.py  (importer module)
    │  └── __init__.py          (Initialize the pysteps_importer_abc package)
    ├── setup.py                (Build and installation script)
    └── MANIFEST.in             (manifest template)

Project name
~~~~~~~~~~~~

::

    pysteps-importer-abc        (project name)

For the project name, our example used the following convention:
**pysteps-importer-<institution short name>**.
Note that this convention is not strictly needed and any name can be used.

Package name
~~~~~~~~~~~~

::

    pysteps-importer-abc
    └── pysteps_importer_abc    (package name)

This is the name of our package containing the new importers for pysteps.
The package name should not contain spaces, hyphens or uppercase letters.
For our example, the package name is **pysteps_importer_abc**.

\__init__.py
~~~~~~~~~~~~

::

    pysteps-importer-abc
        ├── pysteps_importer_abc
        └───── __init__.py

> The __init__.py files are required inform python that a given directory contains a

Reviewer: ```suggestion
The __init__.py files are required to inform python that a given directory contains a
```

python package. Also, this is the first file executed when the importer plugin (i.e.,
the package) is imported.

Importer module
~~~~~~~~~~~~~~~

::

    pysteps-importer-abc
        ├── pysteps_importer_abc
        └───── importer_abc_xyz.py  (importer module)

Inside the package folder (*pysteps_importer_abc*), we place the python module

> (or modules) containing the importers' implementation.

Reviewer: ```suggestion
(or modules) containing the actual implementation of our new importers.
```

The Manifest.in file contains the list of additional files and directories to be
included in your source distribution.

Next, we show an example of a Manifest file that includes a README and the LICENSE files
located in the project root. Lines starting with **#** indicate comments, and they are
ignored.

::

    # This file contains the additional files included in your plugin package

    include LICENSE

    include README.rst

    ###You can also add directories with data, tests, etc.
    # recursive-include dir_with_data

    ###Include the documentation directory, if any.
    # recursive-include doc

For more information about the manifest file, see
https://docs.python.org/3/distutils/sourcedist.html#specifying-the-files-to-distribute

Cookiecutter pysteps-plugin template
------------------------------------

`Cookiecutter`_ is a command-line utility to creates python packages projects from
templates, called "cookiecutters".
To facilitate the creation of an imported plugin for pysteps, the following cookiecutter
template is available at
https://github.com/aperezhortal/cookiecutter-pysteps-plugin
.

> The first step is two install the latest version of `Cookiecutter`_ using,

Reviewer: ```suggestion
The first step is to install the latest version of `Cookiecutter`_ using,
```
